### PR TITLE
A couple of fixes

### PR DESCRIPTION
**Pull request by keithy**
I needed a way to iterate over public properties of the data object.

### CODE REVIEW

**Comment by keithy** [comment]
Why it says 7 commits to merge I do not know!
There are only two changes which I thought were included in my original PR.

I am liking it so far, thanks M.
K.

**Review by mbrowne — APPROVED** [review]

**Comment by mbrowne** [comment]
For getting public data properties, while I don't see any issue with making the `isPublicDataProperty` method protected (otherwise I wouldn't have merged this PR of course), I'm curious as to why this wouldn't work:
`get_object_vars($rolePlayingObject)`

**Comment by mbrowne** [comment]
...oh, maybe you're trying to access them not just inside a Context but inside a role method via `$this`...

**Comment by keithy** [comment]
Hi Matt,

thanks for being so responsive!

On 25 May 2019, at 23:44, Matt Browne wrote:

> ...oh, maybe you're trying to access them not just inside a Context
> but inside a role method via `$this`...
>

Yes exactly, I found that use case within the role methods was not well
supported by built in functions,
and there you had a ready made pre-cached solution.

I have found myself writing a DCI implementation of a database
interface, used like so...

...

     function setUp(): void {
         $this->whenPersisting = new UseDatabase\Sql( /*a PDO Instance*/
$this->setUpPdo(static::environment) );
     }

     // So the test case holds the context.

     function testUseCase_RegisterNewAccountWithSMS_new() {

         // create a model with partial data
         $model = new Account();
         $model->sms = '+1234567890';

         // obtain existing account with this sms if any

         // The db-context gives the model instance a role in this
context,
         // and performs a
UseDatabase\Sql\Roles\NotYetReadModel->fetch()

         $found = $this->whenPersisting->useAsQuery($model)->fetch();

         $this->assertEmpty($found);

         // if no results - build the new account instance
         $model->accounts_id = new Ulid();
         $model->verified = 0;

		 // The db-context gives the model instance a role in this context,
		 // and performs a
UseDatabase\Sql\Roles\NotYetPersistedModel->create()
         $this->persistenceContext->useAsNew($model)->create();

     }
...

I am liking the way this is working out, enabling the Model to be
completely unaware of persistence issues.
However given the way that PHP works, in that everything starts from
source on every invocation there may be no real reason to take the
performance hit of this dynamic implementation of DCI. I haven't tried
any benchmarking so I do not know what the overhead really is.

As it stands this database interface has no method naming conflicts and
so could be provided in alternative ways:

1) as a trait added to the Model abstract class
    (either - manually changing the code for each different db in each
application)
    (or - loading a different db-interface-package via composer)*
2) as a single trait dynamically added to the class, at class load time.
3) a subclass with the trait, that the autoloader loads in lieu of the
standard class.

In other words I thinking that it would be useful to have a simple
implementation, a poor-mans (static) DCI, that conforms to the same
interface, that would really be useful in PHP, retaining similar
architectural boundaries, but facilitate "build/load time" rather than
"run-time" binding.

The simplest idea is that of using composer to choose a trait package.

thoughts?

thanks for this I am enjoying using it so far.

Keith

**Comment by mbrowne** [comment]
Yes, there is some performance overhead with this library, but I doubt it would have a very significant impact unless you're running a website that gets a very large amount of traffic, in which case the database and PHP's synchronous nature are more likely to be the bottlenecks. But I haven't benchmarked it either so I can't say for sure. The rule with performance, of course, is always to measure and see how significant the differences are before changing approaches.

Having said that, this implementation isn't necessarily the "best" way to implement DCI in PHP. A number of DCI implementations for other languages act as transpilers (which we refer to as "source transformation" on the DCI group) which avoids the whole issue of injecting (and later removing) role methods into objects. So those implementations are called "injectionless". So my [example](https://github.com/mbrowne/dci-php/blob/master/examples/TransferMoney/TransferMoney.php) could transpile to something like this:

```php
    class TransferMoney {
        ...

        function transfer() {
            $this->SourceAccount_transferOut($this->amount);
        }
        
        private function SourceAccount_transferOut($amount) {
            $this->SourceAccount_withdraw($amount);
            $this->DestinationAccount_deposit($amount);
        }
        
        private function DestinationAccount_deposit($amount) {
            $this->destinationAccount->increaseBalance($amount);
        }
        
        private function SourceAccount_withdraw($amount) {
            $this->sourceAccount->decreaseBalance($amount);
        }
    }
```

And if you really had time to invest in it, you could even add custom syntax, such as a `role` keyword, and allowing roles to be defined inside of classes (PHP doesn't allow inner classes or traits, but really roles are supposed to belong to the same scope as the context where they're used.)

That would move everything to build time, which would be better for performance of course, but of course it would be a lot of work to write such a transpiler and it would increase the complexity and learning curve for anyone else maintaining your app.

So I think my implementation is a fairly happy medium...yes, there are various compromises you could do, including writing code like the above manually instead of creating a transpiler to do it. But then you're kind of breaking the mental model of DCI, especially for anyone else who maintains your app.

But of course, every situation has to be evaluated independently, and it may well be that in your situation it makes more sense to just use compile-time traits as you would normally do in PHP. Although C++ doesn't have traits, the official C++ implementation of DCI by James Coplien essentially does just that—but C++ has the benefit of a powerful type system that provides a way to discourage the use of a role with an object it wasn't designed for. So if you care about enforcing correct DCI practices, just using traits directly wouldn't be sufficient. But if it's just you maintaining the code for now, that might not be a big concern.

So I'm essentially saying "it's up to you" without giving a real recommendation, but hopefully that gives you some helpful things to think about :)
